src/librbd/cache/pwl/ImageCacheState.cc

```diff
@@ -42,35 +43,33 @@ ImageCacheState<I>::ImageCacheState(I *image_ctx, plugin::Api<I>& plugin_api) :
                             << dendl;
 
   ConfigProxy &config = image_ctx->config;
-  log_periodic_stats = config.get_val<bool>("rbd_persistent_cache_log_periodic_stats");
-  cache_type = config.get_val<std::string>("rbd_persistent_cache_mode");
+  mode = config.get_val<std::string>("rbd_persistent_cache_mode");
 }
 
 template <typename I>
 ImageCacheState<I>::ImageCacheState(
-    I *image_ctx, JSONFormattable &f, plugin::Api<I>& plugin_api) :
+    I *image_ctx, json_spirit::mObject &f, plugin::Api<I>& plugin_api) :
     m_image_ctx(image_ctx), m_plugin_api(plugin_api) {
   ldout(image_ctx->cct, 20) << "Initialize RWL cache state with data from "
                             << "server side"<< dendl;
 
-  present = (bool)f["present"];
-  empty = (bool)f["empty"];
-  clean = (bool)f["clean"];
-  cache_type = f["cache_type"];
-  host = f["pwl_host"];
-  path = f["pwl_path"];
-  uint64_t pwl_size;
-  std::istringstream iss(f["pwl_size"]);
-  iss >> pwl_size;
-  size = pwl_size;
-
-  // Others from config
-  ConfigProxy &config = image_ctx->config;
-  log_periodic_stats = config.get_val<bool>("rbd_persistent_cache_log_periodic_stats");
+  try {
+    present = f["present"].get_bool();
+    empty = f["empty"].get_bool();
+    clean = f["clean"].get_bool();
+    mode = f["mode"].get_str();
+    host = f["pwl_host"].get_str();
+    path = f["pwl_path"].get_str();
+    size = f["pwl_size"].get_uint64();
+  } catch (std::runtime_error& e) {
+    lderr(image_ctx->cct) << "failed to parse persistent cache state: "
+                          << e.what() << dendl;
+  }
 }
 
 template <typename I>
 void ImageCacheState<I>::write_image_cache_state(Context *on_finish) {
+  stats_timestamp = ceph_clock_now();
   std::shared_lock owner_lock{m_image_ctx->owner_lock};
   JSONFormattable f;
   ::encode_json(IMAGE_CACHE_STATE.c_str(), *this, &f);
```

Comment:
I don't know how to revise here.

Reply:
Thanks for so detailed guidance. After discarding `::encode_json` and constucting `json_spirit::mObject` to write, is the `dump() `function still useful?  just keep it there?

Reply:
I folded in the removal of `dump` into the first patch.